### DISCUSSION TITLE
Problem on convert AFP to PDF when AFP have tag BRG

### DISCUSSION THREAD

**Discussion by teeraw-22** [yan74/afplib#70] [Q&A]
I tested to convert simple AFP from your sample code, and convert PDF is work fine.

But when I take AFP full composed from Papyrus Docexec, It will error as this message:
Exception in thread "main" java.lang.ClassCastException: org.afplib.afplib.impl.BRGImpl cannot be cast to org.afplib.afplib.BDT

I seem stucked on BRG (Resource Group)

![image](https://user-images.githubusercontent.com/86512062/182551226-b3c03eb9-64a1-4b80-82ca-da8ace26d0aa.png)

**Comment by yan74** [comment]
Hi @teeraw-22,
I need a little more information. So the error message says your code is expecting a Begin Document (BDT) but is encountering a Begin Resource Group (BRG). This code - is it from the sample directory (in that case which file) or have you written your own code then I would have to see in order to understand.

**Comment by teeraw-22** [comment]
Sorry for some missing information
I used your sample code but changed AFP file. This code will just print out information from AFP file (Still not convert PDF at this time)
Then I was found error as shown previous:

```
try(AfpInputStream ain = new AfpInputStream(new FileInputStream("hello.afp"))) {
			
			BDT bdt = (BDT) ain.readStructuredField();
			EDT edt = (EDT) ain.readStructuredField();
			
			System.out.println(bdt.getDocName());
			System.out.println(bdt.getReserved());
			System.out.println(bdt.getTriplets().size());
			System.out.println(((Comment)bdt.getTriplets().get(0)).getComment());

			System.out.println(edt.getDocName());
			
		} catch (IOException e) {
			e.printStackTrace();
		}
```

**Comment by yan74** [comment]
Ok now I understand. So the Hello World sample code demonstrates how to write a new AFP file and also how to read it. The reading part makes some assumptions on the structure of the AFP, since we know what we have written. Again this is just to demonstrate what can be done. In a more real world scenario you would have to analyze first what type of structured field you loaded. You can have a look at the DumpAFP sample (org.afplib.samples/src/main/java/org/afplib/samples/DumpAFP.java) which is dumping the structure of an AFP file or maybe even better at the org.afplib.samples/src/main/java/org/afplib/samples/CountPagesAndResources.java sample. This shows how to loop over all structured fields https://github.com/yan74/afplib/blob/488750e5aa3b0f71ae578c13470d6c63632da8f4/org.afplib.samples/src/main/java/org/afplib/samples/CountPagesAndResources.java#L57 and how to make decisions based on which type of structured field you encounter https://github.com/yan74/afplib/blob/488750e5aa3b0f71ae578c13470d6c63632da8f4/org.afplib.samples/src/main/java/org/afplib/samples/CountPagesAndResources.java#L60-L71

Once you know the type of a structured field you can cast it to the appropriate type (e.g. BRG brg = (BRG) sf)

Hope that helps,
Yan
